include/quadrilateral_gimp_element.tcc

```diff
@@ -0,0 +1,230 @@
+                                     -1., -3.,
+                                      1., -3.,
+                                      3., -3.,
+                                      3., -1.,
+                                      3.,  1.,
+                                      3.,  3.,
+                                      1.,  3.,
+                                     -1.,  3.,
+                                     -3.,  3.,
+                                     -3.,  1.,
+                                     -3., -1.).finished();
+  // clang-format on
+  return local_nodes;
+}
+
+//! Return shape functions of a 16-node Quadrilateral GIMP Element at a given
+//! local coordinate
+template <>
+inline Eigen::VectorXd mpm::QuadrilateralGIMPElement<2, 16>::shapefn(
+    const Eigen::Matrix<double, 2, 1>& xi, const VectorDim& particle_size,
+    const Eigen::Matrix<double, 2, 1>& deformation_gradient) const {
+
+  //! length of element in local coordinate
+  const double element_length = 2.;
+  //! Nodes in GIMP function
+  const unsigned nfunctions = 16;
+  //! Dimension
+  const unsigned dim = 2;
+  //! Natural nodal coordinates
+  const Eigen::Matrix<double, nfunctions, dim> local_nodes =
+      this->natural_nodal_coordinates();
+  //! To store shape functions
+  Eigen::Matrix<double, nfunctions, 1> shapefn;
+
+  //! loop to iterate over nodes
+  for (unsigned n = 0; n < nfunctions; ++n) {
+    //! local shape function in current plane (x, y or z)
+    Eigen::Matrix<double, dim, 1> sni;
+    //! loop to iterate over dimensions
+    for (unsigned i = 0; i < dim; ++i) {
+      double ni = local_nodes(n, i);
+      double npni = xi(i) - ni;  // local particle  - local node
+      //! Conditional shape function statement see: Bardenhagen 2004
+      if (npni <= (-element_length - particle_size(i))) {
+        sni(i) = 0.;
+      } else if ((-element_length - particle_size(i)) < npni &&
+                 npni <= (-element_length + particle_size(i))) {
+        sni(i) = std::pow(element_length + particle_size(i) + npni, 2.) /
+                 (4. * (element_length * particle_size(i)));
+      } else if ((-element_length + particle_size(i)) < npni &&
+                 npni <= -particle_size(i)) {
+        sni(i) = 1. + (npni / element_length);
+      } else if (-particle_size(i) < npni && npni <= particle_size(i)) {
+        sni(i) = 1. - (((npni * npni) + (particle_size(i) * particle_size(i))) /
+                       (2. * element_length * particle_size(i)));
+      } else if (particle_size(i) < npni &&
+                 npni <= element_length - particle_size(i)) {
+        sni(i) = 1. - (npni / element_length);
+      } else if ((element_length - particle_size(i)) < npni &&
+                 npni <= element_length + particle_size(i)) {
+        sni(i) = std::pow(element_length + particle_size(i) - npni, 2.) /
+                 (4. * element_length * particle_size(i));
+      } else  // if ((element_length + particle_size(i)) < npni)
```

Comment:
Should this else actually check for an error? where you particle is outside of `xp - xv`?